### PR TITLE
[OIDC 8] Evaluate federated credential and generate API key

### PR DESCRIPTION
**Pull request by joelverhagen**
Progress on https://github.com/NuGet/NuGetGallery/issues/10212.
Depends on https://github.com/NuGet/NuGetGallery/pull/10285.

This is the second PR in addition to https://github.com/NuGet/NuGetGallery/pull/10269 which implements the core business logic of the OIDC feature. This PR introduces a high-level `FederatedCredentialService` with a `GenerateApiKeyAsync` method.

Given a NuGet.org username and a bearer token (JWT), this method:
1. Gets federated credential policies for the provided user
2. Evaluates each policy to see if it applies to the bearer token (via a dependency interface)
3. The first matching policy (if any) is used to create short-lived API key.
4. The API key in plaintext is returned to the caller.

A new token exchange HTTP API endpoint will be added in a future PR to call this new method.

Writing out new audit entries will come in a future PR.

This new method uses another class to evaluate the list of polices against the bearer token (which internally includes checking the format and validity of the bearer token):
https://github.com/NuGet/NuGetGallery/blob/4ca98319e435a60807ff8a52735f5da12b68add8/src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialEvaluator.cs#L40

The new `GenerateApiKeyAsync` method is responsible for the following validations:
- The username must refer to a valid, non-deleted, non-locked user account.
- Handling the result of the policy evaluation (bad token, no matching policy, matching policy)
- The package owner on the selected policy must be a valid, non-deleted, non-locked user or organization account.
- The user must have access to the package owner (e.g. member of the organization)
- The federated credential has not already been used (reject token replay)

### CODE REVIEW

File: src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialService.cs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+            // perform validations after the policy evaluation to avoid leaking information about the related users
+
+            var currentUserError = ValidateCurrentUser(currentUser);
+            if (currentUserError != null)
+            {
+                return currentUserError;
+            }
+
+            var packageOwner = _userService.FindByKey(policyEvaluation.MatchedPolicy.PackageOwnerUserKey);
```
Comment (zhhyu) [inline src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialService.cs:88]: >.MatchedPolicy [](http://example.com/codeflow?start=70&length=14)

Will it be helpful to add null check here?

Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialService.cs:88]: This is done inside `ValidatePackageOwner`.

Comment (zhhyu) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialService.cs:88]: Is it possible that `policyEvaluation.MatchedPolicy` is null?

Comment (joelverhagen) [reply src/NuGetGallery.Services/Authentication/Federated/FederatedCredentialService.cs:88]: Oh, I'm sorry, I misread your original message. It is not possible. The property will throw InvalidOperationException if it is null. It is expected that the caller checks the result type first before accessing `MatchedPolicy`.

**Review by zhhyu — APPROVED** [review]

**Review by zhhyu — APPROVED** [review]

**Pull request merged**